### PR TITLE
Fix sprintf() argument count edge cases

### PR DESCRIPTION
**Pull request by staabm**
refs https://github.com/phpstan/phpstan/issues/2342

### CODE REVIEW

File: tests/PHPStan/Rules/Functions/data/printf.php [hunk 1]
```diff
@@ -9,7 +9,7 @@
 sprintf('%2$s %1$s %% %1$s %%%', 'one'); // one parameter missing
 sprintf('%2$s %%'); // two parameters required
 sprintf('%2$s %1$s %1$s %s %s %s %s'); // four parameters required
-sprintf('%2$s %1$s %% %s %s %s %s %%% %%%%', 'one', 'two', 'three', 'four'); // ok
```
Comment (staabm) [inline tests/PHPStan/Rules/Functions/data/printf.php:12]: this test-case was wrong, see https://3v4l.org/2dPI6

the changes in this PR result in properly detecting the error within this test, but only on php8+.

for php 7 the "wrong" test assumption of this test is not changed.
I figured its not worth complicating the already complex regex based parsing (which - as of this PR - is only required for php7 and can be dropped when we raise the min php version to 8).

File: tests/PHPStan/Rules/Functions/data/printf.php [hunk 2]
```diff
@@ -52,3 +52,5 @@
 sprintf($variousPlaceholderCount, 'bar');
 sprintf($variousPlaceholderCount, 'bar', 'baz');
 sprintf($variousPlaceholderCount, 'bar', 'baz', 'lorem');
+
+sprintf('UPDATE {{%loans}} SET field=1');
```
Comment (staabm) [inline tests/PHPStan/Rules/Functions/data/printf.php]: from https://github.com/phpstan/phpstan/issues/2342#issuecomment-516841318

File: src/Rules/Functions/PrintfParametersRule.php [hunk 3]
```diff
@@ -109,6 +112,17 @@ public function processNode(Node $node, Scope $scope): array
 
 	private function getPlaceholdersCount(string $functionName, string $format): int
 	{
+		try {
```
Comment (staabm) [inline src/Rules/Functions/PrintfParametersRule.php:115]: instead of re-implemeting the complex php-src argument parsing, I just delegate to the runtime and interpret the error message.

this works as of php8+ when php-src started throwing `ArgumentCountError`

File: src/Rules/Functions/PrintfParametersRule.php [hunk 4]
```diff
@@ -109,6 +112,17 @@ public function processNode(Node $node, Scope $scope): array
...
+			@sprintf($format);
+		} catch (ArgumentCountError $e) {
+			sscanf($e->getMessage(), '%i arguments are required, %i given', $required, $given);
+
+			if (is_int($required)) {
+				return $required - 1;
+			}
+		}
+
+		// fallback for php < 8
```
Comment (staabm) [inline src/Rules/Functions/PrintfParametersRule.php:129]: the whole regex parsing can be deleted when phpstan min-version is raised to php8+

**Comment by staabm** [comment]
The error is legit https://github.com/PrestaShop/PrestaShop/blob/ca9d81aae0bb17f3e767bb5d835348ed144a3ab5/src/Core/Shop/LogoUploader.php#L112

File: src/Rules/Functions/PrintfParametersRule.php (hunk 4 above)
Comment (VincentLanglet) [reply src/Rules/Functions/PrintfParametersRule.php:129]: Shouldnt an early return be added for php>8 users then if the previous checks are enough then ?

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Please also analyse `printf-php8.php` on PHP 7 in a separate test case method, probably with different errors asserted.

**Comment by ondrejmirtes** [comment]
Does this close https://github.com/phpstan/phpstan/issues/2342? If yes, please change the PR description.

**Comment by staabm** [comment]
> Does this close [phpstan/phpstan#2342](https://github.com/phpstan/phpstan/issues/2342)? If yes, please change the PR description.

it is only fixed for php8+. for php7 the issue still exists. if you are fine with that, I can update the description to a `closes`.

**Comment by ondrejmirtes** [comment]
If PHP 7 still crashes then it's not really fixed...

**Comment by staabm** [comment]
Ok but PR is stll acceptable after I add more tests?

**Comment by ondrejmirtes** [comment]
It signals to me that this is wrong approach. `getPlaceholdersCount` should simply report the correct number, without catching exceptions. So the regexes should be fixed.

**Comment by staabm** [comment]
> It signals to me that this is wrong approach. `getPlaceholdersCount` should simply report the correct number, without catching exceptions. So the regexes should be fixed.

I agree, fixing the regexes would be the best move regarding php version support for this fix.
I tried and couldn't come up with a good regex which works and doesn't regress existing functionality.

I just committed a small fix in [cf6ded3](https://github.com/phpstan/phpstan-src/pull/2041/commits/cf6ded3ab37d1b57084a5b7fad3f7da698d9c042) so you get an idea, that the whole regex thing is only doing php7 stuff after this PR. the newly added 5 lines of code cover all existing logic on php8 (and maybe even cases we did not yet implement - or may be implemented in php-src in the future - because it re-uses the php-src native format checking).

from my point of view this is a big step forward, which allows us to remove a lot of code when phpstan min php-version will be raised to php8+. its also a great way to get rid of clumsy regex stuff which is prone to errors in the future.

I was not able to come up with a regex fix for this problem. I will not invest time in doing more research for such a edge case bug. I am fine with closing the PR so, in case you don't like this php version inconsistency.

**Comment by ondrejmirtes** [comment]
I feel like we should continue with the regex path into the future. I like it more than parsing exception error message which can change at any time.

**Comment by staabm** [comment]
fwiw, this PR also detected https://github.com/phpstan/phpstan/issues/8547
